File: 2018/12.py
with open('12.in') as f:
    input = f.readlines()
initial_state = set(i for i, p in enumerate(input[0][len('initial state: '):]) if p == '#')
rules = {tuple(r[i] == '#' for i in range(5)): r[9] == '#' for r in input[2:]}

def gen_states(prev_states):
    last_state = prev_states[-1]
    next_state = set(i for i in range(min(last_state) - 2, max(last_state) + 3) if rules[tuple(j in last_state for j in range(i - 2, i + 3))])
    for i in range(-1, 2):
        if {j + i for j in last_state} == next_state: return prev_states + [next_state]
    return gen_states(prev_states + [next_state])

states = gen_states([initial_state])
print(sum(states[20]))
shift = 0
for i in range(-1, 2):
    if {j + i for j in states[-2]} == states[-1]:
        shift = i
print(shift)
print(sum(states[-1]) + shift*(50000000000 - len(states) + 1)*len(states[-1]))

File: 2018/12.scala
import scala.annotation.tailrec
import scala.io.Source

val input = Source.fromFile("12.txt").getLines().toSeq
val initial_state = input.head.slice("initial state: ".length, input.head.length).zipWithIndex.collect{case ('#', i) => i}.toSet
implicit val rules: Map[Seq[Boolean], Boolean] = input.slice(2, input.length).map(r => ((0 until 5).map(i => r(i) == '#'), r(9) == '#')).toMap[Seq[Boolean], Boolean]

@tailrec
def gen_states(prev_states: Seq[Set[Int]])(implicit rules: Map[Seq[Boolean], Boolean]): Seq[Set[Int]] = {
  val last_state = prev_states.last
  val next_state = (last_state.min - 2 to last_state.max + 2).filter(i => rules((i - 2 to i + 2).map(last_state))).toSet
  if ((-1 to 1).exists(i => last_state.map(_ + i) == next_state)) prev_states :+ next_state
  else gen_states(prev_states :+ next_state)
}

val states = gen_states(Seq(initial_state))
println(states(20).sum)
val shift = (-1 to 1).filter(i => states(states.length - 2).map(_ + i) == states.last).head
println(states.last.sum + shift*(50000000000L - states.length + 1)*states.last.size)
